great_ape_safe/ape_api/euler.py

```diff
@@ -0,0 +1,48 @@
+from brownie import interface
+from helpers.addresses import r
+
+
+class Euler:
+    def __init__(self, safe):
+        self.safe = safe
+        self.euler = safe.contract(r.euler.euler)
+        self.markets = interface.IEuler(r.euler.euler_markets, owner=safe.account)
+
+    def deposit(self, underlying, amount):
+        etoken = interface.IEToken(
+            self.markets.underlyingToEToken(underlying), owner=self.safe.account
+        )
+
+        before_etoken = etoken.balanceOf(self.safe)
+        before_underlying = underlying.balanceOf(self.safe)
+
+        underlying.approve(self.euler, amount)
+        etoken.deposit(0, amount)
+
+        after_etoken = etoken.balanceOf(self.safe)
+        after_underlying = underlying.balanceOf(self.safe)
+
+        assert after_etoken > before_etoken
+        assert after_underlying < before_underlying
+
+    def deposit_all(self, underlying):
+        self.deposit(underlying, underlying.balanceOf(self.safe))
+
+    def withdraw(self, underlying, amount):
+        etoken = interface.IEToken(
+            self.markets.underlyingToEToken(underlying), owner=self.safe.account
+        )
+
+        before_etoken = etoken.balanceOf(self.safe)
+        before_underlying = underlying.balanceOf(self.safe)
+
+        etoken.withdraw(0, amount)
+
+        after_etoken = etoken.balanceOf(self.safe)
+        after_underlying = underlying.balanceOf(self.safe)
+
+        assert after_etoken < before_etoken
+        assert after_underlying > before_underlying + amount
+
+    def withdraw_all(self, underlying):
+        self.withdraw(underlying, self.safe.balanceOf(underlying))
```

Comment:
i think here the intention is to read the bal of the underlying
```
        self.withdraw(underlying, underlying.balanceOf(self.safe))
```

but it may happen that the underlying balance is way more than the etoken balance in the *etoken*. i think it is worthy perhaps to have a flag in the *withdraw* method which if *True* just to pass as amount *etoken.balanceOf(self.safe)* instead, which should remove the whole position

Reply:
actually no need to query etoken balance in`withdraw_all`, `type(uint).max` ([ref](https://docs.euler.finance/developers/getting-started/contract-reference#withdraw)) can just be passed

Reply:
5906ba5 (also fixes faulty withdraw assert)